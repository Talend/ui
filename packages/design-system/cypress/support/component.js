//github.com/cypress-io/cypress-component-testing-apps/blob/main/react-webpack5-js/cypress/support/component.js
// import './commands';

// Alternatively you can use CommonJS syntax:
// require('./commands')

https: import { mount } from 'cypress/react';
// Ensure global styles are loaded
// import '../../src/index.css';

Cypress.Commands.add('mount', mount);
