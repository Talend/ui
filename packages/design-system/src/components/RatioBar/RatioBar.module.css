/* stylelint-disable color-hex-case */
/* stylelint-disable color-hex-case */
.tc-ratio-bar {
  align-items: center;
  display: flex;
  height: 0.5rem;
  padding: 10px 0;
  width: 100%;
}
.tc-ratio-bar-counter {
  padding-left: 5px;
}
.tc-ratio-bar-lines, .tc-ratio-bar .tc-ratio-bar-line-error, .tc-ratio-bar .tc-ratio-bar-line-filled, .tc-ratio-bar .tc-ratio-bar-line-empty {
  border-radius: 1px;
  height: 100%;
  width: 100%;
}
.tc-ratio-bar .tc-ratio-bar-line {
  height: 0.4rem;
  margin: 0 2px 0 0;
  min-width: 0.4rem;
  transition: var(--coral-transition-instant);
}
.tc-ratio-bar .tc-ratio-bar-line:hover {
  height: 0.5rem;
}
.tc-ratio-bar .tc-ratio-bar-line:first-child {
  margin-left: 0;
}
.tc-ratio-bar .tc-ratio-bar-line:last-child {
  margin-right: 0;
}
.tc-ratio-bar .tc-ratio-bar-line-grow {
  flex-grow: 1;
}
.tc-ratio-bar .tc-ratio-bar-line-empty {
  background-color: var(--coral-color-charts-neutral-weak);
}
.tc-ratio-bar .tc-ratio-bar-line-filled {
  background-color: var(--coral-color-charts-default);
}
.tc-ratio-bar .tc-ratio-bar-line-error {
  background-color: var(--coral-color-charts-danger);
}