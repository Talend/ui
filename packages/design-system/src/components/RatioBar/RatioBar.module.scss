@use '@talend/bootstrap-theme/src/theme/guidelines' as *;
@use '@talend/design-tokens/lib/tokens';

$custom-ratio-bar-height: 0.5rem;
$custom-ratio-bar-line-height: 0.4rem;
$custom-ratio-bar-line-hover-height: 0.5rem;

.tc-ratio-bar {
	align-items: center;
	display: flex;
	height: $custom-ratio-bar-height;
	padding: $padding-small 0;
	width: 100%;

	&-counter {
		padding-left: $padding-smaller;
	}

	&-lines {
		border-radius: 1px;
		height: 100%;
		width: 100%;
	}

	& &-line {
		height: $custom-ratio-bar-line-height;
		margin: 0 2px 0 0;
		min-width: $custom-ratio-bar-line-height;
		transition: tokens.$coral-transition-instant;

		&:hover {
			height: $custom-ratio-bar-line-hover-height;
		}

		&:first-child {
			margin-left: 0;
		}

		&:last-child {
			margin-right: 0;
		}

		&-grow {
			flex-grow: 1;
		}

		&-empty {
			@extend .tc-ratio-bar-lines;
			background-color: tokens.$coral-color-charts-neutral-weak;
		}

		&-filled {
			@extend .tc-ratio-bar-lines;
			background-color: tokens.$coral-color-charts-default;
		}

		&-error {
			@extend .tc-ratio-bar-lines;
			background-color: tokens.$coral-color-charts-danger;
		}
	}
}
