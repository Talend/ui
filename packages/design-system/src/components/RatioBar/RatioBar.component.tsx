import { ReactNode } from 'react';

import { RatioBarComposition } from './RatioBarComposition.component';
import { EmptyLine, ErrorLine, FilledLine } from './RatioBarLines.component';

import styles from './RatioBar.module.scss';

const getFilledValues = (amount: number, total: number) => {
	if (!amount || amount < 0) {
		return { percentage: 0, amount: 0 };
	}

	if (amount > total) {
		return { percentage: 100, amount };
	}

	return { percentage: (amount / total) * 100, amount };
};

const getEmptyValues = (amount: number, total: number) => {
	if (!amount || amount < 0) {
		return { percentage: 100, amount: total };
	}

	if (amount > total) {
		return { percentage: 0, amount: 0 };
	}

	return { percentage: (1 - amount / total) * 100, amount: total - amount };
};

const getLabel = (
	errors: number,
	total: number,
	amount?: number,
	notApplicableLabel?: ReactNode,
) => {
	if (!amount && amount !== 0) {
		return (
			<div className={styles['tc-ratio-bar-counter']} data-testid="ratio-bar-counter">
				{notApplicableLabel || (
					<>
						<strong>N</strong> / A
					</>
				)}
			</div>
		);
	}

	return (
		<div className={styles['tc-ratio-bar-counter']} data-testid="ratio-bar-counter">
			<strong>{amount + errors}</strong>/{total}
		</div>
	);
};

type RatioBarProps = {
	amount?: number;
	errors?: number;
	hideLabel?: boolean;
	notApplicableLabel?: ReactNode;
	total: number;
};

export const RatioBar = ({
	amount,
	errors = 0,
	hideLabel = false,
	notApplicableLabel,
	total,
}: RatioBarProps) => {
	const filled = getFilledValues(amount || 0, total);
	const error = getFilledValues(errors, total);
	const empty = getEmptyValues((amount || 0) + errors, total);

	return (
		<RatioBarComposition>
			<FilledLine percentage={filled.percentage} value={filled.amount} />
			<ErrorLine percentage={error.percentage} value={error.amount} />
			<EmptyLine percentage={empty.percentage} value={empty.amount} />
			{!hideLabel && getLabel(errors, total, amount, notApplicableLabel)}
		</RatioBarComposition>
	);
};
