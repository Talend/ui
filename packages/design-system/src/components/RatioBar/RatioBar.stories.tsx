import { RatioBar as RatioBarComponent } from './RatioBar.component';

export default {
	title: 'Components/Dataviz/RatioBar',
};

export const RatioBar = () => (
	<section style={{ maxWidth: 500, padding: 20 }}>
		<header>Ratio Bar</header>
		<div>
			<div>Not applicable amount</div>
			<RatioBarComponent total={12} />
			<div>With an amount of 0</div>
			<RatioBarComponent amount={0} total={12} />
			<div>With an amount of 10/12</div>
			<RatioBarComponent amount={10} total={12} />
			<div>With an amount of 12/12</div>
			<RatioBarComponent amount={12} total={12} />
			<div>With an amount of 532/1000</div>
			<RatioBarComponent amount={532} total={1000} />
			<div>With an amount of 10/20 with 1 error</div>
			<RatioBarComponent amount={10} errors={1} total={20} />
			<div>With an amount of 532/1000 and no label</div>
			<RatioBarComponent amount={532} total={1000} hideLabel />
		</div>
	</section>
);
