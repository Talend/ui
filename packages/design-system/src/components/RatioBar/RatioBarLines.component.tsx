import styles from './RatioBar.module.css';
import { RatioBarLine } from './RatioBarComposition.component';

type RatioBarLineProps = {
	value: number;
	percentage: number;
};

export const FilledLine = ({ value, percentage }: RatioBarLineProps) => (
	<RatioBarLine
		percentage={percentage}
		value={value}
		className={styles['tc-ratio-bar-line-filled']}
		dataTestId="ratio-bar-filled"
	/>
);

export const EmptyLine = ({ value, percentage }: RatioBarLineProps) => (
	<RatioBarLine
		percentage={percentage}
		value={value}
		className={styles['tc-ratio-bar-line-empty']}
		dataTestId="ratio-bar-empty"
	/>
);

export const ErrorLine = ({ value, percentage }: RatioBarLineProps) => (
	<RatioBarLine
		percentage={percentage}
		value={value}
		className={styles['tc-ratio-bar-line-error']}
		dataTestId="ratio-bar-error"
	/>
);
