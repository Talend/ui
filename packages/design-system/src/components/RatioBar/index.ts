import { RatioBar as RatioBarComponent } from './RatioBar.component';
import { RatioBarComposition, RatioBarLine } from './RatioBarComposition.component';

export type RatioBarType = typeof RatioBarComponent & {
	Composition: typeof RatioBarComposition;
	Line: typeof RatioBarLine;
};

const RatioBar = RatioBarComponent as RatioBarType;
RatioBar.Composition = RatioBarComposition;
RatioBar.Line = RatioBarLine;

export { RatioBarComposition, RatioBarLine };

export default RatioBar;
