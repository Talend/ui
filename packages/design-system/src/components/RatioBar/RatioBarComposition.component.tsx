import { ReactNode } from 'react';

import classNames from 'classnames';

import { Tooltip } from '@talend/design-system/lib/components/Tooltip';

import styles from './RatioBar.module.scss';

const minPercentage = 5;

type RadioBarLineProps = {
	percentage: number;
	value: number;
	tooltipLabel?: string;
	className: string;
	dataTestId?: string;
	dataFeature?: string | null;
	onClick?: any;
};

export function RatioBarLine({
	percentage,
	tooltipLabel,
	className,
	value,
	dataFeature,
	onClick,
	dataTestId,
}: RadioBarLineProps) {
	const canGrow = percentage >= minPercentage;

	if (!value || value < 0) return null;

	function onKeyDown(event: any) {
		switch (event.key) {
			case 'Enter':
				onClick(event);
				break;
			case ' ':
			case 'Spacebar':
				event.preventDefault(); // prevent scroll with space
				event.stopPropagation();
				onClick(event);
				break;
			default:
				break;
		}
	}

	const content = (
		<div
			className={classNames(
				styles['tc-ratio-bar-line'],
				{
					[styles['tc-ratio-bar-line-grow']]: canGrow,
				},
				className,
			)}
			tabIndex={0}
			style={{
				flexBasis: `${Math.max(percentage, minPercentage)}%`,
			}}
			data-testid={dataTestId}
			data-test={dataTestId}
			role={onClick && 'button'}
			data-feature={dataFeature}
			onClick={onClick}
			onKeyDown={onKeyDown}
		>
			{tooltipLabel && <span className="sr-only">{tooltipLabel}</span>}
		</div>
	);

	if (!tooltipLabel) {
		return content;
	}

	return (
		<Tooltip title={tooltipLabel} placement="bottom">
			{content}
		</Tooltip>
	);
}

export const RatioBarComposition = ({ children }: { children: ReactNode }) => (
	<div className={styles['tc-ratio-bar']} data-testid="ratio-bar">
		{children}
	</div>
);
