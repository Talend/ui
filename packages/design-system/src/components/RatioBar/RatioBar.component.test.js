import { describe, expect, it } from '@jest/globals';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { RatioBar } from './RatioBar.component';

describe('RatioBar', () => {
	describe('RatioBar component', () => {
		it('should render an empty bar', () => {
			// given
			const props = {
				amount: 0,
				total: 12,
			};
			// when
			const { container } = render(<RatioBar {...props} />);
			// then
			expect(container.firstChild).toMatchSnapshot();
			expect(screen.getByTestId('ratio-bar-counter')).toHaveTextContent('0/12');
			expect(screen.getByTestId('ratio-bar')).toBeVisible();
			expect(screen.getByTestId('ratio-bar-empty')).toBeVisible();
			expect(screen.getByTestId('ratio-bar-empty')).toHaveStyle('flex-basis: 100%');
		});

		it('should render a not applicable chart', () => {
			// given
			const props = {
				amount: null,
				total: 12,
			};
			// when
			render(<RatioBar {...props} />);
			// then
			expect(screen.getByTestId('ratio-bar-counter')).toHaveTextContent('N/A');
		});

		it('should render an full sized chart', async () => {
			const user = userEvent.setup();

			// given
			const props = {
				amount: 12,
				total: 12,
			};
			// when
			render(<RatioBar {...props} />);
			// then
			await user.tab();
			expect(screen.getByTestId('ratio-bar-counter')).toHaveTextContent('12/12');
			expect(screen.getByTestId('ratio-bar-filled')).toHaveStyle('flex-basis: 100%');
		});

		it('should render a classic ratio bar', () => {
			// given
			const props = {
				amount: 5,
				total: 12,
			};
			// when
			render(<RatioBar {...props} />);
			// then
			expect(screen.getByTestId('ratio-bar-filled')).toHaveStyle('flex-basis: 41.66666666666667%');
			expect(screen.getByTestId('ratio-bar-empty')).toHaveStyle('flex-basis: 58.33333333333333%');
			expect(screen.getByTestId('ratio-bar-counter')).toHaveTextContent('5/12');
		});
	});

	it('should render a classic ratio bar with errors', () => {
		// given
		const props = {
			amount: 5,
			total: 12,
			errors: 2,
		};
		// when
		render(<RatioBar {...props} />);
		// then
		expect(screen.getByTestId('ratio-bar-filled')).toHaveStyle('flex-basis: 41.66666666666667%');
		expect(screen.getByTestId('ratio-bar-empty')).toHaveStyle('flex-basis: 41.666666666666664%');
		expect(screen.getByTestId('ratio-bar-error')).toHaveStyle('flex-basis: 16.666666666666664%');
		expect(screen.getByTestId('ratio-bar-counter')).toHaveTextContent('7/12');
	});

	it('should render a classic ratio bar without label', () => {
		// given
		const props = {
			amount: 5,
			total: 12,
		};
		// when
		render(<RatioBar {...props} hideLabel />);
		screen;
		// then
		expect(screen.getByTestId('ratio-bar-filled')).toHaveStyle('flex-basis: 41.66666666666667%');
		expect(screen.getByTestId('ratio-bar-empty')).toHaveStyle('flex-basis: 58.33333333333333%');
		expect(screen.queryByTestId('ratio-bar-counter')).not.toBeInTheDocument();
	});
});
