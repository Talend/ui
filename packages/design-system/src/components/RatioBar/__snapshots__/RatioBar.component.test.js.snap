// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RatioBar RatioBar component should render an empty bar 1`] = `
<div
  class="theme-tc-ratio-bar"
  data-testid="ratio-bar"
>
  <div
    class="theme-tc-ratio-bar-line theme-tc-ratio-bar-line-grow theme-tc-ratio-bar-line-empty"
    data-test="ratio-bar-empty"
    data-testid="ratio-bar-empty"
    style="flex-basis: 100%;"
    tabindex="0"
  />
  <div
    class="theme-tc-ratio-bar-counter"
    data-testid="ratio-bar-counter"
  >
    <strong>
      0
    </strong>
    /
    12
  </div>
</div>
`;
